game.py

```diff
@@ -0,0 +1,261 @@
+import winningpath as w
+import player as p
+import node as n
+import random
+import time
+from subprocess import call
+
+class Game:
+
+    def __init__(self, player1, player2, boardsize):
+        self.player1 = player1
+        self.player2 = player2
+        self.boardsize = boardsize
+        self.winner = None
+        self.winningpaths = []
```

Comment:
I think there's a few parts of the computer's move selection that could be refactored to make the execution flow a little simpler to understand:

1. `BoardState` Model
   - Right now there is no explicit object that models a board *state* (ie the grid and location of pieces on it)
   - This is the most important refactor to make; it will simplify the control flow as well as the logic of computing moves & winning 
   - This should look something like this:

```
class BoardState(object):
   """
   Put all abstractions and logic for a board state in here; a sample skeleton API is below.
   """
    def __init__(self, ...):
      pass 
    def is_move_valid(self, move):
      pass # return True or False for a given move on this state
    def is_move_winning(self, move):
      pass # return True or False for a given move on this state
    def valid_moves(self):
      # ... compute some stuff
      return moves # some iterable of all valid moves that could be made on this state
    def transition(self, move):
      # ... compute some stuff
      return type(self)(...) # return a *new* State that has been altered by the provided (valid) move
``` 

2. Take all of the board-related logic out of the `Game`

- The `Game` object should just be a way of *coordinating* state transitions for players, but there shouldn't be any logic for *what* moves should be made; a game should call other objects to make those choices
- The contol flow will simplify quite a bit here; it would look a bit more like this:
```
class Game(object):
   DEFAULT_BOARDSIZE = (3, 3)
   def __init__(self, players, init_state=None):
       self.players = players # store players as a list---2 players is the same as N players after all...  ;) 
       self.state = init_state or State.from_boardsize(DEFAULT_BOARD_SIZE) # create a state object
       self.init_board = board

   def iterate(self):
      # Put some kind of iteration in which 1 player moves and the self.state transitions
      # A "game" is really just a sequence of states, so the playing logic should be a sequence of calls to iterate
      # Since there are up to len(players) per turn, this could all happen here, or iterate could call another private function or method 
 
      return self   # return a pointer to this object, which now has an updated state

   def main(self):
      # this method should basically have no code lol
      while not self.state.is_winning():
          self.iterate()
```

3. Handle human & computer players in the `Player` class
  - A player should have a method like `choose_move(self, state)` that returns a move for a board
  - For humans, this is a call to get user input
  - For computers, this is where the board moves should be evaluated
  - This suggests that there should be a `Player` base class and then a `HumanPlayer` and `ComputerPlayer` subclasses that override the `choose_move(self, state)` as appropriate

Reply:
4. Use a 2D data storage container (matrix/array) as the backing for a `BoardState`

- The `BoardState` is an abstraction for a board but it still needs an actual variable backing it
- For this you can use `list` of `list`s or also an `n` by `m` `numpy.array` (https://numpy.org/doc/stable/reference/generated/numpy.array.html) to store the contents of each `(x,y)` location in the board
- I'm a big fan of `numpy` since the array object has a really powerful API to express mathematical row & column operations, so that you can manipulate the board representation very succinctly
- The later advantages of this are that you can some of the properties of the board matrix to then calculate whether a state is winning (e.g. a vertical/horizontal win can be determined from row-wise & column-wise operations, and diagonal win on a board can be determined from row/col-wise operations on the board state after a [*rotation matrix*](https://en.wikipedia.org/wiki/Rotation_matrix#In_two_dimensions) has been applied to the board). It will also be a really good format of data to use for any form of optimization of which move to make for a given board, e.g. with [linear programming](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html). The suite of `numpy`/`scipy`/`matplotlib` libraries is the "python scientific computing stack" and really really powerful. Using an array here for the data structure (and then also trying some `scipy` optimization routines for choosing the best move would be a great learning experience!